File: cpp/comm/radio.cpp
#include "comm/radio.hpp"
#include <iostream>
using std::vector;
namespace comm {
    Radio::Radio(serial::Serial * serialIn, RadioConfig & configIn) :
        mode(RADIO_OFF),
        config(configIn),
        serial(serialIn)
    {
        rxBuffer.reserve(config.rxBufferSize);
    };

    bool Radio::setMode(RadioMode modeIn) {
        bool retValue = false;
        switch (modeIn) {
            case RADIO_OFF:
                retValue = setOff();
                break;
            case RADIO_SLEEP:
                retValue = setSleep();
                break;
            case RADIO_RECEIVE:
                retValue = setReceive();
                break;
            case RADIO_TRANSMIT:
                retValue = setTransmit();
                break;
        }

        return retValue;
    }

    bool Radio::setOff() {
        if (mode == RADIO_OFF) {
            return false;
        }
        else {
            mode = RADIO_OFF;
            return true;
        }
    }
    
    bool Radio::setSleep() {
        if (mode == RADIO_SLEEP) {
            return false;
        }
        else {
            mode = RADIO_SLEEP;
            return true;
        }
    }
    
    bool Radio::setReceive() {
        if (mode == RADIO_RECEIVE) {
            return false;
        }
        else {
            mode = RADIO_RECEIVE;
            return true;
        }
    }

    bool Radio::setTransmit() {
        if (mode == RADIO_TRANSMIT) {
            return false;
        }
        else {
            mode = RADIO_TRANSMIT;
            return true;
        }
    }
    
    void Radio::clearRxBuffer() {
        rxBuffer.clear();
    }
           
    int Radio::readBytes(bool bufferFlag) {
        // Attempt to read serial bytes
        vector<uint8_t> newBytes;
        serial->read(newBytes, config.numBytesToRead);
       
        // Process received bytes
        if (newBytes.size() > 0) {
            processRxBytes(newBytes, bufferFlag);
        }

        return newBytes.size();
    }
            
    void Radio::sendBytes(std::vector<uint8_t> msgBytes) {
        serial->write(msgBytes);
    }


    int Radio::processRxBytes(vector<uint8_t> & newBytes, bool bufferFlag) {
        // Base class just adds received bytes to the buffer.
        return bufferRxMsg(newBytes, bufferFlag);
    }

    int Radio::bufferRxMsg(vector<uint8_t> & newBytes, bool bufferFlag) {
        if (newBytes.size() == 0) {
            return 0;
        }
        
        if (bufferFlag == true) { // Buffer new bytes
            if ((rxBuffer.size() + newBytes.size()) <= config.rxBufferSize) {
                rxBuffer.insert(rxBuffer.end(), newBytes.begin(), newBytes.end());
                return newBytes.size();
            }
            else { // Buffer full so discard bytes
                return -1;
            }
        }
        else { // Replace any existing bytes
            rxBuffer = newBytes;
            return newBytes.size();
        }   

    }

    vector<uint8_t> Radio::getRxBytes(void) {
        return rxBuffer;
    }

    unsigned int Radio::sendMsg(vector<uint8_t> & msgBytes) {
        if (msgBytes.size() > 0) {
            return serial->write(createMsg(msgBytes));
        }

        return 0; // no data sent
    }
    
    void Radio::bufferTxMsg(vector<uint8_t> & msgBytes) {
        txBuffer.insert(txBuffer.end(), msgBytes.begin(), msgBytes.end());
    }

    unsigned int Radio::sendMsg(Message & msg) {
        return 0;
    }

    vector<uint8_t> Radio::createMsg(vector<uint8_t> msgBytes) {
        // Base class just passes through raw bytes
        return msgBytes;
    }

    unsigned int Radio::sendBuffer(unsigned int maxBytesToSend) {
        if (txBuffer.size() > 0) {
            if (maxBytesToSend > 0 && txBuffer.size() > maxBytesToSend) { // Too much data to send
                vector<uint8_t> msgBytes = vector<uint8_t>(txBuffer.begin(), txBuffer.begin() + maxBytesToSend);
                sendMsg(msgBytes);
                txBuffer.erase(txBuffer.begin(), txBuffer.begin() + maxBytesToSend); // clear sent bytes
                return 1; // partial buffer sent
            }
            else { // send entire tx buffer
                sendMsg(txBuffer);
                txBuffer.clear();
            }
        }

        return 0;
    }

    void Radio::sendCommand(std::vector<uint8_t> cmd) {
        
    }
}

File: python/mesh/generic/radio.py
from enum import IntEnum
from switch import switch
import serial
from mesh.generic.customExceptions import NoSerialConnection

class RadioMode(IntEnum):
    off = 0
    sleep = 1
    receive = 2
    transmit = 3

class Radio(object):

    def __init__(self, serial, config):
        self.serial = serial
        self.mode = RadioMode.off
        
        # Read
        self.uartNumBytesToRead = config['uartNumBytesToRead']
        self.rxBufferSize = config['rxBufferSize']
        self.clearRxBuffer()

        # Send
        self.txBuffer = bytearray()
    
    def setMode(self, mode):
        """Change radio operating mode (i.e. rx, tx, sleep, off)."""
        if mode != self.mode:
            for case in switch(mode):
                if case(RadioMode.off):
                    self.setOff()
                    break
                if case(RadioMode.sleep):
                    self.setSleep()
                    break
                if case(RadioMode.receive):
                    self.setReceive()
                    break
                if case(RadioMode.transmit):
                    self.setTransmit()
                    break

    def setSleep(self):
        self.mode = RadioMode.sleep

    def setOff(self):
        self.mode = RadioMode.off

    def setReceive(self):
        self.mode = RadioMode.receive

    def setTransmit(self):
        self.mode = RadioMode.transmit

    # Read methods
    def clearRxBuffer(self):
        """Reset receive buffer."""
        self.rxBuffer = bytearray(self.rxBufferSize)
        self.bytesInRxBuffer = 0

    def readBytes(self, bufferFlag):
        """Reads raw bytes from serial connection"""
        if not self.serial:
            raise NoSerialConnection("No serial connection available")
            return  
        
        # Read from serial port
        newBytes = []
        try:
            newBytes = self.serial.read(self.uartNumBytesToRead)
        except serial.SerialException:
            pass

        # Process rx bytes
        if newBytes:
            self.processRxBytes(newBytes, bufferFlag)

    def sendBytes(self, msgBytes):
        """Send bytes over serial connection."""
        if not self.serial:
            raise NoSerialConnection("No serial connection available")
            return  
        
        try:
            self.serial.write(msgBytes)
        except serial.SerialException:
            pass
            
    def processRxBytes(self, newBytes, bufferFlag):
        """Default behavior just adds directly to rxBuffer."""
        self.bufferRxMsg(newBytes, bufferFlag)

    def bufferRxMsg(self, newBytes, bufferFlag):
        # Put read bytes in buffer
        if not newBytes:
            return

        numNewBytes = len(newBytes)    
        if bufferFlag == True: # Add new bytes read to buffer
            if (self.bytesInRxBuffer + numNewBytes) <= len(self.rxBuffer): # Prevent rx buffer overload  
                self.rxBuffer[self.bytesInRxBuffer:self.bytesInRxBuffer+numNewBytes] = newBytes
                self.bytesInRxBuffer += numNewBytes
        else: # Replace buffer contents with new bytes
            self.rxBuffer[0:numNewBytes] = newBytes
            self.bytesInRxBuffer = numNewBytes

    def getRxBytes(self):
        return self.rxBuffer[0:self.bytesInRxBuffer]

    # Send methods
    def sendMsg(self, msgBytes):
        """Send message to radio."""
        if len(msgBytes) > 0:
            msg = self.createMsg(msgBytes)
            self.sendBytes(msg)

    def bufferTxMsg(self, msgBytes):
        self.txBuffer += msgBytes

    def createMsg(self, msgBytes):
        """Default behavior is to just pass through raw bytes."""
        return msgBytes

    def sendBuffer(self, maxBytesToSend=0):
        if self.txBuffer:
            if maxBytesToSend > 0 and len(self.txBuffer) > maxBytesToSend: # too much data to send
                self.sendMsg(self.txBuffer[:maxBytesToSend])
                self.txBuffer = self.txBuffer[maxBytesToSend:] # remove sent bytes from buffer
            else: # send entire buffer
                self.sendMsg(self.txBuffer)
            
                # Clear tx buffer
                self.txBuffer = bytearray()

    def sendCommand(self, cmd):
        """Issue command to radio."""
        pass
